Adafruit_CircuitPython_74HC595: Added support for multiple shift registers

davidleedom opened the pull request:
I simply changed the single byte used to send to the shift register into an array. Then mapped pin values greater than 7 to the next byte.   So you will see value of _byte_pos which is the element of the byte array and _byte_pin which is the bit in the _byte_pos element.

davidleedom committed:
Adafruit_CircuitPython_74HC595: Corrected pull request issues.

tannewt commented on adafruit_74hc595.py:
```diff
@@ -44,6 +44,8 @@ def __init__(self, pin_number, shift_register_74hc595):
         ShiftRegister74HC595 instance.
         """
         self._pin = pin_number
+        self._byte_pos = int(self._pin / 8)
```
This will do truncated division:
```suggestion
        self._byte_pos = self._pin // 8
```

tannewt commented on adafruit_74hc595.py:
```diff
@@ -100,23 +109,35 @@ def pull(self, val):  # pylint: disable=no-self-use
+        self._number_of_shift_registers = number_of_shift_registers
+        self._gpio = bytearray(self._number_of_shift_registers)
+        for x in range(0, self._number_of_shift_registers - 1):
+            self._gpio[x] = 0x00
```
bytearrays start with zeroes.
```suggestion
```

tannewt commented on adafruit_74hc595.py:
```diff
@@ -100,23 +109,35 @@ def pull(self, val):  # pylint: disable=no-self-use
     @gpio.setter
     def gpio(self, val):
-        self._gpio[0] = val & 0xFF
+
+        for x in range(0, self._number_of_shift_registers - 1):
+            self._gpio[x] = val[x] & 0xFF
```
```suggestion
        self._gpio = val
```

No need to mask because bytearrays will only allow 0-255. (You will want to check that the val type is bytes or bytearray though.)

tannewt reviewed (changes requested):
Thank you for this! Just a couple suggestions. We'll want to release this as a new version number because `gpio` will return a bytearray instead of an int now.

davidleedom committed:
Update adafruit_74hc595.py

I'm new to Python.  Makes sense to me.

Co-authored-by: Scott Shawcroft <scott@tannewt.org>

davidleedom committed:
Update adafruit_74hc595.py

I did not know this.

Co-authored-by: Scott Shawcroft <scott@tannewt.org>

davidleedom committed:
Update adafruit_74hc595.py

I didn't really understand what this was doing in the original code, but I thought I'd just keep it.

Co-authored-by: Scott Shawcroft <scott@tannewt.org>

tannewt commented:
Did you test the latest version? I don't always make perfect edits...

davidleedom reviewed (commented):
I have accepted all your changes.  I also applied them to my local copy and tested them in my project which is using four shift registers.  It all appears to work just fine.

davidleedom commented:
I think I completed everything, but being this is more first Pull Request experience I might be missing something.

davidleedom commented:
Yes i did.

On Tue, Apr 13, 2021, 2:07 PM Scott Shawcroft ***@***.***>
wrote:

> Did you test the latest version? I don't always make perfect edits...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/adafruit/Adafruit_CircuitPython_74HC595/pull/17#issuecomment-818947744>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFBGG5DE76HLE3WQSLPIWETTISB63ANCNFSM42X2MRAQ>
> .
>

tannewt commented:
> I have accepted all your changes. I also applied them to my local copy and tested them in my project which is using four shift registers. It all appears to work just fine.

That's exactly what I had hoped. Thanks!

tannewt reviewed (approved):
Thank you!
